eventmesh-sdks/eventmesh-sdk-java/src/main/java/org/apache/eventmesh/client/grpc/consumer/SubStreamHandler.java

```diff
@@ -129,6 +132,12 @@ public void run() {
     }
 
     public void close() {
+        // Avoid closing multiple times
+        if (isClosed) {
+            return;
+        }
+        isClosed = true;
```

Comment:
The way you add the flag doesn't solve the problem of multiple closing under concurrency.

Reply:
The `onCompleted` method of `StreamObserver` is idempotent. After the first call to the `onCompleted` method, the stream is considered to be completed, and any subsequent `onNext` or `onCompleted` calls will be ignored.

Since the `sender == null` statement has been removed, so I reverted the close flag in https://github.com/apache/eventmesh/pull/4807/commits/878009a99b1b64ffdec0b579bb564b8c9e31afec.

Reply:
> The `onCompleted` method of `StreamObserver` is idempotent. After the first call to the `onCompleted` method, the stream is considered to be completed, and any subsequent `onNext` or `onCompleted` calls will be ignored.

Are you sure that grpc just ignores subsequent calls rather than throwing an exception that could be and should be avoided?
> Since the `sender == null` statement has been removed, so I reverted the close flag in [878009a](https://github.com/apache/eventmesh/commit/878009a99b1b64ffdec0b579bb564b8c9e31afec).

The removal of `sender == null` is not the main reason for reverting this newly added close flag. The main reason is that this close flag was used in a wrong way for concurrent programming.

Reply:
Sorry, `onCompleted` doesn't throw an exception, `onNext` does. I forgot to remove `onNext` after pasting the text.

There's only `onCompleted` call here, which should not throw an exception if `onCompleted` itself didn't throw one.

Is there anything I should change?

Reply:
> `onCompleted` doesn't throw an exception (when called again)

Could you please provide the basis for this conclusion? Was it based on testing with multiple calls, or some other evidence? Because from the grpc source code, it seems that it cannot explain why a secondary calling `onCompleted` would not throw an exception.

Reply:
https://github.com/grpc/grpc-java/blob/497e1552172632107574d67d7c092a3f5a39a980/stub/src/main/java/io/grpc/stub/ServerCalls.java#L393-L397

When `onCompleted` is called, it will set `completed` flag to true.

https://github.com/grpc/grpc-java/blob/497e1552172632107574d67d7c092a3f5a39a980/stub/src/main/java/io/grpc/stub/ServerCalls.java#L359-L381

When `onNext` is called, it will check `completed` flag and throw an exception.

Anyway, I added a check in https://github.com/apache/eventmesh/pull/4807/commits/8e812cdcb68a7a1e90241b30739cba77998b82e5 to keep `senderOnComplete` called once.

Reply:
> https://github.com/grpc/grpc-java/blob/497e1552172632107574d67d7c092a3f5a39a980/stub/src/main/java/io/grpc/stub/ServerCalls.java#L393-L397
> 
> When `onCompleted` is called, it will set `completed` flag to true, but won't throw exceptions.

I suggest you take another look inside `call.close(Status.OK, new Metadata())` line of this piece of source code.

Reply:
Thanks for your reminding, it will throw an exception.